client/src/app/interceptors/auth.interceptor.ts

```diff
@@ -0,0 +1,73 @@
+import { Injectable } from '@angular/core';
+import {
+  HttpRequest,
+  HttpHandler,
+  HttpEvent,
+  HttpInterceptor,
+  HttpClient
+} from '@angular/common/http';
+import { environment } from 'src/environments/environment';
+import { catchError, Observable, throwError } from 'rxjs';
+import { User } from '../models/user.model'
+import { Token } from '../interfaces/token';
+
+
+/**
+ * Interceptor for HTTP requests, it attaches Authorization 
+ * Header and attempts to do a token refresh when  current 
+ * access Token expires. 
+ * At other layers ensure user is redirected to login 
+ * if Token expired error is thrown.
+ */
+@Injectable()
+export class AuthInterceptor implements HttpInterceptor {
+  private  MAXRETRIES = 2;
+  private retries = 0;
+
+
+  constructor(private http: HttpClient) { }
+
+  intercept(request: HttpRequest<unknown>, next: HttpHandler): Observable<HttpEvent<unknown>> {
+
+    let accessToken = '';
+    const localUser = localStorage.getItem('user');
+    if (localUser) {
+      const user = User.fromJson(localUser);
```

Comment:
Using `auth.user` will make the call to `fromJson` unnecessary. Also: We already have a function `User().deserialize()`, please pull the latest code

Reply:
I would like to have deserialise as a static function that returns user object. 
I didnt want to change that function hence i introduces fromJSon as in intermediate step to reduce blast radius of changes

Reply:
That would leave us with two functions basically doing the same thing. I think we should agree on one approach and stick to that.

I actually do like the static approach and parsing JSON within the static function. Reduces redundant code when using it.